File: iOS/Sources/Presentation/ViewModels/BudgetListViewModel.swift
import Foundation
import Combine

@MainActor
class BudgetListViewModel: ObservableObject {
    @Published var budgets: [Budget] = []
    @Published var summary: BudgetSummary?
    @Published var currentPeriod: String
    @Published var errorMessage: String?
    @Published var isLoading = false

    private let budgetInteractor: BudgetInteractor

    init(budgetInteractor: BudgetInteractor) {
        self.budgetInteractor = budgetInteractor
        self.currentPeriod = BudgetInteractor.currentPeriod()
    }

    func loadBudgets() async {
        isLoading = true
        do {
            budgets = try await budgetInteractor.getBudgets(forPeriod: currentPeriod)
            summary = try await budgetInteractor.getSummary(forPeriod: currentPeriod)
        } catch {
            errorMessage = error.localizedDescription
        }
        isLoading = false
    }

    func deleteBudget(at offsets: IndexSet) async {
        for index in offsets {
            let budget = budgets[index]
            do {
                try await budgetInteractor.deleteBudget(id: budget.id)
                budgets.remove(at: index)
                summary = try await budgetInteractor.getSummary(forPeriod: currentPeriod)
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}

File: android/app/src/main/java/com/budgetmanager/ui/viewmodel/BudgetListViewModel.kt
package com.budgetmanager.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.budgetmanager.domain.entity.Budget
import com.budgetmanager.domain.entity.BudgetSummary
import com.budgetmanager.domain.interactor.BudgetInteractor
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class BudgetListState(
    val budgets: List<Budget> = emptyList(),
    val summary: BudgetSummary? = null,
    val currentPeriod: String = BudgetInteractor.currentPeriod(),
    val isLoading: Boolean = false,
    val errorMessage: String? = null
)

class BudgetListViewModel(
    private val budgetInteractor: BudgetInteractor
) : ViewModel() {

    private val _state = MutableStateFlow(BudgetListState())
    val state: StateFlow<BudgetListState> = _state.asStateFlow()

    fun loadBudgets() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, errorMessage = null)
            try {
                val period = _state.value.currentPeriod
                val budgets = budgetInteractor.getBudgetsByPeriod(period)
                val summary = budgetInteractor.getSummary(period)
                _state.value = _state.value.copy(
                    budgets = budgets,
                    summary = summary,
                    isLoading = false
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isLoading = false,
                    errorMessage = e.message ?: "Failed to load budgets"
                )
            }
        }
    }

    fun deleteBudget(id: Long) {
        viewModelScope.launch {
            try {
                budgetInteractor.deleteBudget(id)
                loadBudgets()
            } catch (e: Exception) {
                _state.value = _state.value.copy(errorMessage = e.message)
            }
        }
    }

    fun clearError() {
        _state.value = _state.value.copy(errorMessage = null)
    }
}
